File: clients/android/app/src/main/java/eu/allodia/mailcal/AllodiaSubscriptionCard.kt
// Settings → Allodia account → Subscription: what is being charged, by whom, until when, and the
// way to start one. Its Apple twin is AllodiaSubscriptionSettings.swift; Windows and Linux do not
// draw this yet. Keep the wording in step.
//
// The rules are `purchasing.md`, the contract beside the Allodia Licence, and the core holds every
// one of them. **What this file draws is what the core answered.** Which offers exist and in what
// order is `orderAllodiaOffers`, and a store's subscription is changed only at that store, so the
// button opens `manageUrl` rather than doing anything itself.
//
// ⚠️ **A price is drawn, never parsed and never assembled.** Play requires the store's own
// formatted price be shown, the store's commission is already inside it, and the core carries no
// locale data to format one with.
package eu.allodia.mailcal

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Card
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import java.util.Locale
import uniffi.mailcal_bindings.AllodiaOffer
import uniffi.mailcal_bindings.AllodiaPlan
import uniffi.mailcal_bindings.AllodiaStore
import uniffi.mailcal_bindings.AllodiaStoreSubscription
import uniffi.mailcal_bindings.AllodiaSubscription

/**
 * What the card needs, beyond the read itself.
 *
 * Held by the activity rather than the composable because both of its flows leave the app: a Play
 * sheet is another process and Allodia's checkout is a browser, so state this screen owned would
 * not still be here when the person came back.
 *
 * [accountId] is the Allodia account a store purchase would be tagged with, absent on a device
 * that signed in before the id was recorded. [buying] is the period whose sheet is up, so that
 * button alone spins. [note] is the last thing worth saying about an attempt.
 */
internal data class AllodiaSubscriptionUi(
    val state: AllodiaSubscriptionState = AllodiaSubscriptionState.Checking,
    val accountId: String? = null,
    val buying: AllodiaPlan? = null,
    val note: String? = null,
)

/**
 * Drawn only while somebody is signed in: a subscription belongs to an account, and there is
 * nothing to say about one that does not exist.
 */
@Composable
internal fun AllodiaSubscriptionCard(
    ui: AllodiaSubscriptionUi,
    onRefresh: () -> Unit,
    onBuy: (AllodiaPlan) -> Unit,
    onManageStore: (AllodiaStoreSubscription) -> Unit,
    onSignInAgain: () -> Unit,
    onClosed: () -> Unit,
    onCancel: () -> Unit,
    onResubscribe: () -> Unit,
    onSwitch: (AllodiaPlan) -> Unit,
) {
    val ctx = LocalContext.current
    // The read runs when the card appears rather than when the activity connects: it is a network
    // round trip, and a category nobody opened is not worth one.
    LaunchedEffect(Unit) { onRefresh() }
    // ⚠️ **A flow the shop never answers would leave every button disabled behind a spinner that
    // will not stop**, and leaving the card is then the only way back. Letting go of the wait is
    // not abandoning the purchase: Play reports it on its own listener, which starts a pass of its
    // own, and an unattached purchase is refunded rather than lost.
    DisposableEffect(Unit) { onDispose { onClosed() } }
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.fillMaxWidth().padding(16.dp)) {
            Text(
                L10n.settings_subscription_heading(ctx),
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(bottom = 8.dp),
            )
            when (val state = ui.state) {
                AllodiaSubscriptionState.Checking ->
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        CircularProgressIndicator(modifier = Modifier.size(16.dp))
                        Secondary(L10n.settings_subscription_checking(ctx))
                    }
                // Deliberately quiet. A read that did not arrive costs a sentence, never access:
                // whether a capability is on is `allodiaEntitlement`'s answer, and it is local.
                AllodiaSubscriptionState.Unavailable ->
                    Secondary(L10n.settings_subscription_unavailable(ctx))
                // An offer rather than an error: they are signed in and this one read is asleep,
                // and the ordinary sign-in asks for the full current scope set.
                AllodiaSubscriptionState.NeedsReauth -> {
                    Secondary(L10n.settings_subscription_reauth(ctx))
                    TextButton(onClick = onSignInAgain) {
                        Text(L10n.settings_allodia_reauth_action(ctx))
                    }
                }
                is AllodiaSubscriptionState.Loaded ->
                    Loaded(
                        state.view,
                        ui,
                        onBuy,
                        onManageStore,
                        onSignInAgain,
                        onCancel,
                        onResubscribe,
                        onSwitch,
                    )
            }
            ui.note?.let { note -> Secondary(note, modifier = Modifier.padding(top = 8.dp)) }
        }
    }
}

@Composable
private fun Loaded(
    view: AllodiaSubscriptionView,
    ui: AllodiaSubscriptionUi,
    onBuy: (AllodiaPlan) -> Unit,
    onManageStore: (AllodiaStoreSubscription) -> Unit,
    onSignInAgain: () -> Unit,
    onCancel: () -> Unit,
    onResubscribe: () -> Unit,
    onSwitch: (AllodiaPlan) -> Unit,
) {
    val ctx = LocalContext.current
    if (view.subscription.entitled) {
        Active(view.subscription, onManageStore)
        // What may be done to Allodia's own subscription, which the service decided
        // (AllodiaSubscriptionWrites.kt). A store's is changed at that store and nothing here
        // touches it.
        AllodiaSubscriptionWrites(view.subscription, onCancel, onResubscribe, onSwitch)
    } else {
        Secondary(L10n.settings_subscription_free(ctx))
        if (allodiaNeedsReauthToBuy(view.offers, ui.accountId)) {
            Secondary(L10n.settings_subscription_reauth(ctx))
            TextButton(onClick = onSignInAgain) {
                Text(L10n.settings_allodia_reauth_action(ctx))
            }
        }
        Offers(allodiaBuyableOffers(view.offers, ui.accountId), ui.buying, onBuy)
    }
    // ⚠️ Money was taken and nothing has been granted. A card that stays silent here leaves
    // somebody with no way to find that out.
    if (view.anythingStuck) Secondary(L10n.settings_subscription_stuck(ctx))
    // ⚠️ The other ending that owes somebody an explanation: what they bought is on a different
    // Allodia account, so it is never coming to this one and no amount of waiting changes that.
    if (view.claimedElsewhere) Secondary(L10n.settings_subscription_claimed_elsewhere(ctx))
}

/** What a paid account says: who is charging, until when, and where to change it. */
@Composable
private fun Active(
    subscription: AllodiaSubscription,
    onManageStore: (AllodiaStoreSubscription) -> Unit,
) {
    val ctx = LocalContext.current
    val locale: Locale = allodiaReaderLocale()
    val billers = allodiaBillers(subscription)
    billers.firstOrNull()?.let { first ->
        Text(
            L10n.settings_subscription_billed_by(ctx, allodiaBillerName(first)),
            style = MaterialTheme.typography.bodyMedium,
        )
    }
    subscription.currentPeriodEnd?.let { raw ->
        allodiaDate(raw, locale)?.let { end ->
            Secondary(
                if (allodiaWillRenew(subscription)) {
                    L10n.settings_subscription_renews(ctx, end)
                } else {
                    L10n.settings_subscription_ends(ctx, end)
                }
            )
        }
    }
    // Not a lapse, and never drawn as one: the store is retrying and access continues.
    if (allodiaInGrace(subscription)) {
        billers.firstOrNull()?.let { first ->
            Secondary(L10n.settings_subscription_grace(ctx, allodiaBillerName(first)))
        }
    }
    // Reported, never resolved. Cancelling one of them without asking is a decision about somebody
    // else's money, so each exit is offered and none is taken.
    if (subscription.duplicateBilling.isNotEmpty()) {
        Text(
            L10n.settings_subscription_duplicate(
                ctx,
                subscription.duplicateBilling.joinToString(", ", transform = ::allodiaBillerName),
            ),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.error,
        )
    }
    // A store's subscription is the store's to change, so this opens its page rather than offering
    // a cancel button that would have nothing to call. Review-blocking on both stores, which is why
    // it is drawn for every store that still has something to do rather than for a recognised one.
    allodiaManageableStores(subscription).forEach { store ->
        val biller = if (store.source == AllodiaStore.APPLE) "Apple" else "Google Play"
        TextButton(onClick = { onManageStore(store) }) {
            Text(L10n.settings_subscription_manage(ctx, biller))
        }
    }
}

/** The plans that can be bought here, in the core's order. */
@Composable
private fun Offers(offers: List<AllodiaOffer>, buying: AllodiaPlan?, onBuy: (AllodiaPlan) -> Unit) {
    val ctx = LocalContext.current
    offers.forEach { offer ->
        TextButton(onClick = { onBuy(offer.plan) }, enabled = buying == null) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                // The label stays while the spinner runs. A button that becomes a bare spinner
                // stops saying what it is doing, which is the wrong half to drop when what it is
                // doing is charging somebody.
                if (buying == offer.plan) CircularProgressIndicator(modifier = Modifier.size(16.dp))
                // **The period is on the button, not under it.** It is the thing being chosen, so
                // a row of buttons that all read "Subscribe" makes the reader pair each one with a
                // line of small print to find out what it does.
                Text(
                    when (offer.plan) {
                        AllodiaPlan.YEARLY ->
                            L10n.settings_subscription_buy_yearly(ctx, offer.displayPrice)
                        AllodiaPlan.MONTHLY ->
                            L10n.settings_subscription_buy_monthly(ctx, offer.displayPrice)
                    }
                )
            }
        }
        // What renewing means, beside the button and not a tap away: both stores make this
        // review-blocking, and somebody agreeing to a recurring charge is owed it whether or not
        // they are.
        Secondary(L10n.settings_subscription_terms(ctx))
    }
}

// The locale this reader is actually being shown, which is the configuration's rather than the JVM
// default: the app's own language picker writes the configuration.
@Composable
internal fun allodiaReaderLocale(): Locale =
    LocalConfiguration.current.locales.takeIf { !it.isEmpty }?.get(0) ?: Locale.getDefault()

@Composable
private fun Secondary(text: String, modifier: Modifier = Modifier) {
    Text(
        text,
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = modifier,
    )
}

File: clients/windows/Mailcal/Services/AllodiaSubscriptionCard.cs
// What the subscription card has to draw, and the reading of a core answer that decides it.
//
// The rules are `purchasing.md`, the contract that ships beside the Allodia Licence, and the core
// holds every one of them. What is here is the few things the card can be in, plus the decisions
// that follow from one answer. Its twins are AllodiaSubscriptionModel.kt on Android and
// MailcalModel.AllodiaSubscription.swift on Apple: keep the states and the wording in step.
//
// ⚠️ **None of this gates a capability.** `AllodiaEntitlement` does, locally and without a network
// call. This is what the card *says*, so a service that cannot be reached costs somebody a
// sentence, never access they have paid for.
//
// WinUI-free AND L10n-free on purpose, the same seam SearchHorizonLine and AttendeeSummary use:
// every word arrives as a parameter or is a company's own name, so Mailcal.Tests can link this
// file and fail on the rules without a Windows host. Which matters here more than most: the
// sentence somebody reads about being charged twice, and whether a cancel button exists at all,
// are both invisible to a screenshot and expensive to get wrong.

using System.Collections.Generic;
using System.Linq;
using uniffi.mailcal_bindings;

namespace Allodia.Mailcal.Services;

/// <summary>What the subscription card has to draw.</summary>
internal enum AllodiaSubscriptionStatus
{
    /// <summary>
    /// The read is in flight. Distinct from <see cref="Unavailable"/>: nothing has been learned
    /// yet, and a spinner and a failure are different things to put in front of somebody.
    /// </summary>
    Checking,

    /// <summary>The read did not come back, or this deployment has no billing configured.</summary>
    Unavailable,

    /// <summary>
    /// The read could not be made at all, because this device's sign-in predates the permission it
    /// needs. An offer rather than an error: they are signed in, one thing is asleep, and the
    /// ordinary sign-in asks for the full current scope set. Kept apart from
    /// <see cref="Unavailable"/> because the remedies differ: waiting fixes an outage and never
    /// fixes this.
    /// </summary>
    NeedsReauth,

    /// <summary>The service answered.</summary>
    Loaded,
}

/// <summary>
/// The card's whole state: which of the four above, and the answer when it is
/// <see cref="AllodiaSubscriptionStatus.Loaded"/>.
/// </summary>
/// <remarks>
/// There is no "purchase outstanding" half here, unlike Apple's and Android's. Windows ships no
/// store purchase at all (`purchasing.md`, per-platform status), so nothing is ever left on this
/// device to attach, and a card drawing "a purchase has not gone through yet" would be describing
/// a thing that cannot happen here.
/// </remarks>
internal sealed record AllodiaSubscriptionCardState(
    AllodiaSubscriptionStatus Status,
    AllodiaSubscription? Subscription = null,
    /// <summary>
    /// The period whose checkout is being opened, so that button alone shows the spinner, or
    /// <c>null</c>.
    /// </summary>
    AllodiaPlan? Buying = null,
    /// <summary>The last thing worth saying about an attempt, already localised, or <c>null</c>.</summary>
    string? Note = null);

/// <summary>The decisions the card makes from one read, each pure so each is a test.</summary>
internal static class AllodiaSubscriptionCard
{
    /// <summary>
    /// What a read that did not come back is allowed to put on screen.
    /// </summary>
    /// <remarks>
    /// ⚠️ <b>The core's typed answer decides, never the failure's text.</b> A sign-in that predates
    /// the permission this read needs is an offer with a remedy, and drawing it as an outage tells
    /// somebody to wait for something that will never arrive. The account-list card next door
    /// reads the same answer for the same reason.
    /// </remarks>
    internal static AllodiaSubscriptionStatus ReadFailure(AllodiaGrantHealth health) =>
        health == AllodiaGrantHealth.NeedsReauth
            ? AllodiaSubscriptionStatus.NeedsReauth
            : AllodiaSubscriptionStatus.Unavailable;

    /// <summary>
    /// What a biller is called on screen.
    /// </summary>
    /// <remarks>
    /// <b>Allodia, never the payment processor behind it.</b> Naming a processor somebody has never
    /// heard of, inside a message about being charged twice, is how a correct warning reads as a
    /// scam. The bare company name is right here, where the sentence is about who is taking the
    /// money rather than about the app.
    /// </remarks>
    internal static string BillerName(AllodiaBiller biller) => biller switch
    {
        AllodiaBiller.Apple => "Apple",
        AllodiaBiller.Google => "Google Play",
        // A biller this build cannot name still has to appear, or a "you are being charged twice"
        // warning would name only one of the two.
        AllodiaBiller.Unknown unknown => unknown.Label,
        _ => "Allodia",
    };

    /// <summary>
    /// Everyone charging for this account right now, in a stable order: Allodia's own billing
    /// first, then each store in the order the service listed them.
    /// </summary>
    /// <remarks>
    /// ⚠️ <b>"Right now" is the whole of it, and reading every store the service listed was a bug
    /// on the other clients.</b> The list keeps a subscription after it ends, so an account that
    /// bought at one store and later at another carries both, and naming the first named the dead
    /// one.
    /// <para>
    /// Whether a store still has something to manage is a different question, answered by
    /// <see cref="StoreCanBeManaged"/>, and the two part company on the states that matter most.
    /// </para>
    /// </remarks>
    internal static IReadOnlyList<AllodiaBiller> Billers(AllodiaSubscription subscription)
    {
        var billers = new List<AllodiaBiller>();
        if (subscription.Own is { } own && own.Status is not AllodiaOwnStatus.PendingFirstPayment)
        {
            billers.Add(new AllodiaBiller.Allodia());
        }
        // Distinct for the same reason the manage buttons are: two subscriptions at one store is
        // an ordinary shape, and naming that store twice would say somebody is charged twice by it.
        billers.AddRange(subscription.Stores
            .Where(store => StoreIsBilling(store.Status))
            .DistinctBy(store => store.Source)
            .Select(store => store.Source switch
            {
                AllodiaStore.Apple => (AllodiaBiller)new AllodiaBiller.Apple(),
                AllodiaStore.Google => new AllodiaBiller.Google(),
                // The service never reports its own billing as a store, so this arm is unreachable
                // rather than meaningful. Saying "Allodia" is still the right answer if it is
                // reached.
                _ => new AllodiaBiller.Allodia(),
            }));
        return billers;
    }

    /// <summary>
    /// Whether a store subscription is one somebody is still on: renewing, being retried, or
    /// cancelled and running out the period already paid for.
    /// </summary>
    /// <remarks>
    /// The rest grant nothing, and the status is deliberately read arm by arm rather than by
    /// exclusion: a status this build does not know is <b>never read as permission</b>, which is
    /// the rule the core states about the same enum.
    /// </remarks>
    internal static bool StoreIsBilling(AllodiaStoreStatus status) => status
        is AllodiaStoreStatus.Active or AllodiaStoreStatus.Grace or AllodiaStoreStatus.Cancelled;

    /// <summary>
    /// The stores worth offering a way into, one entry per store rather than per subscription.
    /// </summary>
    /// <remarks>
    /// ⚠️ <b>An account can carry more than one subscription at the same store</b>, and it does the
    /// moment somebody resubscribes: the lapsed one is still inside the period it was paid for, so
    /// both are manageable and both were drawn, as two identical buttons opening the same page. A
    /// store's subscription page is the store's, not the subscription's, so one button is the
    /// whole of what there is to offer.
    /// </remarks>
    internal static IReadOnlyList<AllodiaStoreSubscription> ManageableStores(
        AllodiaSubscription subscription) =>
        subscription.Stores
            .Where(store => StoreCanBeManaged(store.Status))
            .DistinctBy(store => store.Source)
            .ToList();

    /// <summary>
    /// Whether the store still has something for this person to do about this subscription.
    /// </summary>
    /// <remarks>
    /// ⚠️ <b>Not the same question as who is billing them, and the two answers differ on exactly
    /// the states somebody needs most.</b> On hold and paused grant nothing, so neither may claim
    /// the "billed by" line, and both are fixed only at the store: a card that failed is replaced
    /// there and a pause is lifted there. Dropping their button strands the person it matters to.
    /// <para>
    /// Expired and revoked are the other way about. Nothing is left to manage, and offering the
    /// route anyway walks somebody into the store's own resubscribe button while another source is
    /// already charging them, which is the duplicate billing this contract warns about rather than
    /// causes.
    /// </para>
    /// </remarks>
    internal static bool StoreCanBeManaged(AllodiaStoreStatus status) => status
        is AllodiaStoreStatus.Active
        or AllodiaStoreStatus.Grace
        or AllodiaStoreStatus.Cancelled
        or AllodiaStoreStatus.OnHold
        or AllodiaStoreStatus.Paused;

    /// <summary>
    /// Whether anything will charge again, which decides between "renews on" and "runs until".
    /// </summary>
    /// <remarks>
    /// A store subscription says so itself; Allodia's own says so by still having a next payment.
    /// Deliberately not a reading of <c>Entitled</c>, which answers a different question.
    /// </remarks>
    internal static bool WillRenew(AllodiaSubscription subscription)
    {
        if (subscription.Stores.Any(store => store.AutoRenewing))
        {
            return true;
        }
        return subscription.Own is { } own
            && own.Status is AllodiaOwnStatus.Active
            && !string.IsNullOrEmpty(own.NextPaymentDate);
    }

    /// <summary>
    /// Whether a charge has failed and is being retried. <b>Access continues</b>, so this is never
    /// drawn as a lapse.
    /// </summary>
    internal static bool InGrace(AllodiaSubscription subscription) =>
        subscription.Stores.Any(store => store.Status is AllodiaStoreStatus.Grace)
        || subscription.Own?.Status is AllodiaOwnStatus.PastDue;

    /// <summary>
    /// The period a switch would move to, which is simply the other one, or <c>null</c> when there
    /// is nothing to switch: no subscription of Allodia's own, or a service that has said this
    /// account may not switch.
    /// </summary>
    /// <remarks>
    /// <c>Actions</c> decides, and nothing else. The service computed it with every biller in
    /// view, which no client is in a position to do: somebody the App Store is charging has a
    /// period, and switching it through this API is exactly what the service refuses.
    /// </remarks>
    internal static AllodiaPlan? SwitchTarget(AllodiaSubscription subscription)
    {
        if (!subscription.Actions.CanSwitchInterval)
        {
            return null;
        }
        return subscription.Own?.Interval switch
        {
            AllodiaPlan.Monthly => AllodiaPlan.Yearly,
            AllodiaPlan.Yearly => AllodiaPlan.Monthly,
            _ => null,
        };
    }

    /// <summary>
    /// Whether to offer starting the subscription again, which is a different question from
    /// whether it has been cancelled: the service decides, because a cancelled subscription whose
    /// period has run out is a new checkout rather than a restart, and only it knows which.
    /// </summary>
    internal static bool OffersRestart(AllodiaSubscription subscription) =>
        subscription.Actions.CanResubscribe
        && subscription.Own?.Status is AllodiaOwnStatus.Cancelled;

    /// <summary>
    /// The day both confirmations talk about: what has already been paid for, as the service sent
    /// it, or the empty string when it sent none.
    /// </summary>
    /// <remarks>
    /// ⚠️ The date is the whole reassurance. Somebody cancelling wants to know they are not losing
    /// what they have already paid for, and a cancellation with no date reads as "it stops now",
    /// which is the one thing it does not do.
    /// </remarks>
    internal static string WriteDate(AllodiaSubscription subscription) =>
        subscription.Own?.CurrentPeriodEnd ?? subscription.CurrentPeriodEnd ?? string.Empty;
}
